enterprise_gateway/services/processproxies/yarn.py

```diff
@@ -108,49 +111,67 @@ def launch_process(self, kernel_cmd, **kwargs):
         """
+        env_dict = kwargs.get('env', {})
 
-        executer_memory = int(kwargs['env']['KERNEL_EXECUTOR_MEMORY'])
-        driver_memory = int(kwargs['env']['KERNEL_DRIVER_MEMORY'])
-
-        candidate_queue_name = str(kwargs['env']['KERNEL_QUEUE'])
+        executor_memory = int(env_dict.get('KERNEL_EXECUTOR_MEMORY', 0))
+        driver_memory = int(env_dict.get('KERNEL_DRIVER_MEMORY', 0))
 
-        node_label = str(kwargs['env']['KERNEL_NODE_LABEL'])
+        if executor_memory*driver_memory > 0:
+            container_memory = self.resource_mgr.cluster_node_container_memory()
+            if max(executor_memory, driver_memory) > container_memory:
+                self.log_and_raise(http_status_code=500,
+                                   reason="Container Memory not sufficient for a executor/driver allocation")
 
-        self.container_memory = self.resource_mgr.cluster_node_container_memory()
-        self.candidate_queue = self.resource_mgr.cluster_scheduler_queue(candidate_queue_name)
-        self.candidate_partition = self.resource_mgr.cluster_queue_partition(self.candidate_queue, node_label)
+        candidate_queue_name = (env_dict.get('KERNEL_QUEUE', None))
+        node_label = env_dict.get('KERNEL_NODE_LABEL', None)
 
-        if max(executer_memory, driver_memory) > self.container_memory:
-            self.log_and_raise(http_status_code=500,
-                               reason="Container Memory not sufficient for a executor/driver allocation")
+        if candidate_queue_name is None or node_label is None:
+            return
 
         # else the resources may or may not be available now. it may be possible that if we wait then the resources
         # become available. start  a timeout process
 
         self.start_time = RemoteProcessProxy.get_current_time()
+        self.candidate_queue = self.resource_mgr.cluster_scheduler_queue(candidate_queue_name)
+        self.candidate_partition = self.resource_mgr.cluster_queue_partition(self.candidate_queue, node_label)
+
         yarn_available = False
 
-        self.log.info("Yarn Endpoint to check for resources : " + self.yarn_endpoint)
-        self.log.info("Retrying for " + str(self.yarn_resource_check_wait_time) + " ms if resources not found ")
+        self.log.debug("Yarn Endpoint to check for resources : " + self.yarn_endpoint)
+        self.log.debug("Retrying for " + str(self.yarn_resource_check_wait_time) + " ms if resources not found ")
         while not yarn_available:
             self.handle_yarn_queue_timeout()
+            self.log.debug("Checking if " + self.candidate_partition + " has used capacity < 95% ")
             yarn_available = self.resource_mgr.cluster_scheduler_queue_availability(self.candidate_partition)
+            self.log.debug("Result: " + str(yarn_available))
+            # each time we sleep we substract equal amount of time for the kernel launch timeout
+            self.kernel_launch_timeout -= poll_interval
```

Comment:
Let's remove any debug statements from within the loop as these will be posted every half second and nothing is changing.  We can move the first statement before the loop and the fact that we didn't throw an exception implies the result was successful.  (Btw, you might want to keep this in mind when looking into the yarn-api-client implementations since some will be called frequently in loops like this.)

Where does the 'capacity < 95%' come from?  That sounds like a detail of checking the availability and I'm not sure we should be assuming the 95% threshold.  Perhaps we can make that more generic?

My previous comment regarding decrementing `self.kernel_launch_timeout` was to do something like this...
```python
while not yarn_available:
            self.handle_yarn_queue_timeout()
            yarn_available = self.resource_mgr.cluster_scheduler_queue_availability(self.candidate_partition)

check_availability_total_time = RemoteProcessProxy.get_time_diff(self.start_time, RemoteProcessProxy.get_current_time())
self.kernel_launch_timeout -= check_availability_total_time
```
If you'd rather decrement on each iteration, then please move it back into the timeout handler.  I think it would be good to isolate the use of `poll_interval` to as few places as possible.

Reply:
> Where does the 'capacity < 95%' come from? That sounds like a detail of checking the availability and I'm not sure we should be assuming the 95% threshold. Perhaps we can make that more generic?

Sure.
A thing to note here is, right now, we have not implemented the logic of checking the memory requests during kernel launch with the amount of memory available in yarn queue. 
Rather we are checking if the queue is more than 95% full. If yes, we don't allow the kernel to be launched. This logic appeared more generic to us, given that we can actually tune this threshold according to yarn configurations and admin requirements to allow for kernel launch. 
And yes, we should be making this check more generic enough. 
I Will be committing files with relevant changes.

Reply:
Uploaded.